src/server/services/validate-upload.js

```diff
@@ -128,60 +122,118 @@ async function validateSubrecipientRecord ({ upload, recipient, rules, trns }) {
+      await updateRecipient(existing.id, { record: recipient }, trns)
+    } else {
+      const dbRow = {
+        uei: recipient.Unique_Entity_Identifier__c,
+        tin: recipient.EIN__c,
+        record: recipient,
+        upload_id: upload.id
       }
+      await createRecipient(dbRow, trns)
     }
+  }
 
-    // if it's valid, we can insert it into the db
-    if (errors.length === 0) {
-      if (existing?.upload_id === upload.id) {
-        await updateRecipient(existing.id, { record: recipient }, trns)
-      } else {
-        const dbRow = {
-          uei: recipient.Unique_Entity_Identifier__c,
-          tin: recipient.EIN__c,
-          record: recipient,
-          upload_id: upload.id
-        }
-        await createRecipient(dbRow, trns)
+  return errors
+}
+
+async function validateRecord ({ upload, record, typeRules: rules, trns }) {
+  // start by trimming any whitespace
+  for (const key of Object.keys(rules)) {
+    if (record[key] && (typeof record[key]) === 'string') {
+      record[key] = record[key].trim()
+    }
+  }
+
+  // placeholder for rule errors we're going to find
+  const errors = []
+
+  // check all the rules
+  for (const [key, rule] of Object.entries(rules)) {
+    // if the rule only applies on different EC codes, skip it
+    if (rule.ecCodes && (!upload.ec_code || rule.ecCodes.indexOf(upload.ec_code) < 0)) {
+      continue
+    }
+
+    // if there's something in the field, make sure it meets requirements
+    if (record[key]) {
+      // make sure pick value is one of pick list values
+      if (rule.listVals.length > 0 && rule.listVals.indexOf(record[key]) < 0) {
+        errors.push(new ValidationError(
+          `Value for ${key} must be one of ${rule.listVals.length} options in the input template`,
+          { col: rule.columnName, severity: 'err' }
+        ))
+      }
+
+      // make sure max length is not too long
+      if (rule.maxLength && String(record[key]).length > rule.maxLength) {
+        errors.push(new ValidationError(
+          `Value for ${key} cannot be longer than ${rule.maxLength} (currently, ${String(record[key]).length})`,
+          { col: rule.columnName, severity: 'err' }
+        ))
+      }
+
+    // if the field is unset, is that okay?
+    } else {
+      // make sure required keys are present
+      if (rule.required === true) {
+        errors.push(new ValidationError(
+          `Value is required for ${key}`,
+          { col: rule.columnName, severity: 'err' }
+        ))
       }
     }
   }
 
+  // return all the found errors
   return errors
 }
 
-async function validateSubrecipients ({ upload, records, rules, trns }) {
+async function validateRules ({ upload, records, rules, trns }) {
   const errors = []
 
-  // validate each, and save the errors
-  const recipients = records.filter(rec => rec.type === 'subrecipient').map(r => r.content)
+  // go through every rule type we have
+  for (const [type, typeRules] of Object.entries(rules)) {
```

Comment:
In my head it feels more natural to iterate through records first and then find all rules that apply for that record.

I guess it's functionally identical though?  🤷  Feel free to ignore this.

Reply:
it's not, because of [here](https://github.com/usdigitalresponse/arpa-reporter/blob/d80772bf5fb799f3c5d383110a20096b37c5094e/src/server/services/validate-upload.js#L232). we assume that if we filter records by type, the order in which they appear will be the order they appear in their original sheet. if we want to go through records first, and then rules, then we should save the row in the input spreadsheet in the record entry